### PR TITLE
Adding initial fusion test cases for resnet50

### PR DESCRIPTION
**Pull request by manupak**
This commit adds 2 tests cases that occurs in resnet50. Moreover, there is a compilation discrepency between clone and nonclone flows.

### CODE REVIEW

**Comment by manupak** [comment]
The intention of this PR is to just to discuss about the compilation discrepancy in clone vs nonclone flows for some cases.

File: mlir/test/fusion/e2e/resnet50/mixr-resnet-fusion-case-1.clone.mlir [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// RUN: rocmlir-opt -migraphx-to-tosa %s | rocmlir-driver -host-pipeline partition,highlevel -targets %arch | rocmlir-gen -ph -print-results -rand 4 -rand_type float -fut test --verifier clone - | rocmlir-driver -host-pipeline xmodel,runner -kernel-pipeline full -targets %arch | mlir-cpu-runner -O2 --shared-libs=%linalg_test_lib_dir/libmlir_rocm_runtime%shlibext,%conv_validation_wrapper_library_dir/libconv-validation-wrappers%shlibext,%linalg_test_lib_dir/libmlir_runner_utils%shlibext,%linalg_test_lib_dir/libmlir_c_runner_utils%shlibext --entry-point-result=void | FileCheck %s --check-prefix=CLONE
```
Comment (manupak) [inline mlir/test/fusion/e2e/resnet50/mixr-resnet-fusion-case-1.clone.mlir:1]: This one passes

File: mlir/test/fusion/e2e/resnet50/mixr-resnet-fusion-case-2.clone.mlir [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// RUN: rocmlir-opt -migraphx-to-tosa %s | rocmlir-driver -host-pipeline highlevel | rocmlir-gen -ph -print-results -rand none - | rocmlir-driver -arch %arch -c  | mlir-cpu-runner -O2 --shared-libs=%linalg_test_lib_dir/libmlir_rocm_runtime%shlibext,%conv_validation_wrapper_library_dir/libconv-validation-wrappers%shlibext,%linalg_test_lib_dir/libmlir_runner_utils%shlibext --entry-point-result=void | FileCheck %s
```
Comment (manupak) [inline mlir/test/fusion/e2e/resnet50/mixr-resnet-fusion-case-2.clone.mlir:1]: This command line passes.

File: mlir/test/fusion/e2e/resnet50/mixr-resnet-fusion-case-2.clone.mlir [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+// RUN: rocmlir-opt -migraphx-to-tosa %s | rocmlir-driver -host-pipeline highlevel -targets %arch | rocmlir-gen -ph -print-results -rand 4 -rand_type float -fut func_rock --verifier clone - | rocmlir-driver -host-pipeline xmodel,runner -kernel-pipeline full -targets %arch | mlir-cpu-runner -O2 --shared-libs=%linalg_test_lib_dir/libmlir_rocm_runtime%shlibext,%conv_validation_wrapper_library_dir/libconv-validation-wrappers%shlibext,%linalg_test_lib_dir/libmlir_runner_utils%shlibext,%linalg_test_lib_dir/libmlir_c_runner_utils%shlibext --entry-point-result=void | FileCheck %s --check-prefix=CLONE
```
Comment (manupak) [inline mlir/test/fusion/e2e/resnet50/mixr-resnet-fusion-case-2.clone.mlir:2]: This command line fails compilation -- which is exactly same as case-1's command line (except for fut)

File: mlir/test/fusion/e2e/resnet50/mixr-resnet-fusion-case-2.nonclone.mlir [hunk 4]
```diff
@@ -0,0 +1,30 @@
+  // RUN: rocmlir-opt -migraphx-to-tosa %s | rocmlir-driver -host-pipeline highlevel | rocmlir-gen -ph -print-results -rand none - | rocmlir-driver -arch %arch -c | mlir-cpu-runner -O2 --shared-libs=%linalg_test_lib_dir/libmlir_rocm_runtime%shlibext,%conv_validation_wrapper_library_dir/libconv-validation-wrappers%shlibext,%linalg_test_lib_dir/libmlir_runner_utils%shlibext,%linalg_test_lib_dir/libmlir_c_runner_utils%shlibext --entry-point-result=void | FileCheck %s
```
Comment (manupak) [inline mlir/test/fusion/e2e/resnet50/mixr-resnet-fusion-case-2.nonclone.mlir:1]: This is my bespoke verification method pre-clone era and it works.
This is a test case migraphx succeeds in compilation.

**Review by manupak — COMMENTED** [review]
Hi @pcf000,

I am trying to understand a compilation failure that occurs in some cases when using the clone functionality. In this PR, I have put three test cases : 

1) this one works and uses clone functionality.
2) this one does not work with clone functionality (maybe Im doing something silly in CL) 
3) this is my bespoke verification method used in pre-clone era but is identical to 2) and it seems to work (compile).

PTAL, when you have some time...

**Comment by manupak** [comment]
Also ./bin/rocmlir-opt -migraphx-to-tosa /home/mkarunar/work/llvm-project-mlir/mlir/test/fusion/e2e/resnet50/mixr-resnet-fusion-case-2.clone.mlir | ./bin/rocmlir-driver -host-pipeline partition,highlevel -targets gfx908:sramecc+:xnack- | ./bin/rocmlir-gen -ph -print-results -rand 4 -rand_type float -fut func_rock --verifier clone - > case-2-rocmlir-gen-output.mlir

[case-2-rocmlir-gen-output.mlir.txt](https://github.com/ROCmSoftwarePlatform/rocMLIR/files/9953896/case-2-rocmlir-gen-output.mlir.txt)

**Comment by manupak** [comment]
The IR does not seem to contain rock.* operations instead linalg.* are there in the func_rock function (as opposed to func_rock_cloned). Therefore, Im worried whether we are testing against pure linalg and not our kernel generation.

**Comment by manupak** [comment]
On the otherhand : the non-cloned one : 

./bin/rocmlir-opt -migraphx-to-tosa /home/mkarunar/work/llvm-project-mlir/mlir/test/fusion/e2e/resnet50/mixr-resnet-fusion-case-2.nonclone.mlir | ./bin/rocmlir-driver -host-pipeline highlevel | ./bin/rocmlir-gen -ph -print-results -rand none - > case-2-noncloned-output.mlir


[case-2-noncloned-rocmlir-gen-output.mlir.txt](https://github.com/ROCmSoftwarePlatform/rocMLIR/files/9954012/case-2-noncloned-rocmlir-gen-output.mlir.txt)

**Comment by manupak** [comment]
Ok found the reason : It is because TosaOptionalDecompositions will convert 1x1 conv to tosa.fully_connected then it becomes opaque to TosaToRock. Therefore, our fusion infra is not exercised at all.

**Comment by pcf000** [comment]
[AMD Official Use Only]

    Date: Monday, November  7, 2022  10:00AM -08
    From: Manupa Karunaratne ***@***.***>

    Ok found the reason : It is because TosaOptionalDecompositions will
    convert 1x1 conv to tosa.fully_connected then it becomes opaque to
    TosaToRock. Therefore, our fusion infra is not exercised at all.

Hadn't looked yet (busy morning), but I suspected something like that.
The clone-verification code depends on the exact shape produced by
cloning kernels -- nested modules for kernels, compiled to binary, with
the original form left behind in the top-level module.

Cloning kernels typically depends on the tosa-partition pass, directly
because it makes the outlined functions and indirectly because those
outlined functions have the "kernel" attribute which the clone-kernels
pass requires.

It's possible to set the tosa-partition pass's anchor ops from the
command line with
"--tosa-partition=anchor-ops=tosa.conv2d,tosa.fully_connected" with
either rocmlir-driver or rocmlir-opt, though it might take some
shuffling to get it to happen at the right time.

**Comment by manupak** [comment]
Thanks @pcf000 for the explanation and I dont think we support tosa.fully_connected yet.
So feels a bit premature to support that anyway,

**Pull request closed without merging**